docs/en/latest/contribute.md

```diff
@@ -21,23 +21,107 @@ title: Contributing to apisix-ingress-controller
+```shell
+cd /path/to/apisix-ingress-controller
+make build
+./apisix-ingress-controller version
+```
+
+## Test
+
+### Run unit test cases
+
+```shell
+cd /path/to/apisix-ingress-controller
+make unit-test
+```
+
+### Run e2e test cases
+
+We using [KIND](https://kind.sigs.k8s.io/) for running e2e test cases. Please ensure `kind` CLI has been installed.
+Currently we using KIND latest version v0.11.1 and using Kubernetes v1.21.1 for testing.
+
+```shell
+cd /path/to/apisix-ingress-controller
+make e2e-test
+```
+
+Note the running of e2e cases is somewhat slow, so please be patient.
+
+> See [here](https://onsi.github.io/ginkgo/#focused-specs) to learn
+how to just run partial e2e cases.
+
+### Build docker image
+
+```shell
+cd /path/to/apisix-ingress-controller
+make build-image IMAGE_TAG=a.b.c
+```
+
+> Note: The Dockerfile in this repository is only for development, not for release.
+
+If you're coding for apisix-ingress-controller and adding some e2e test cases to verify your changes,
+you should push the images to the image registry that your Kubernetes cluster can access, if you're using Kind, just run the following command:
+
+```shell
+make push-images-to-kind
+```
+
+## Run apisix-ingress-controller locally
+
+We assume all prerequisites above mentioned are met, what's more, since we want to run apisix-ingress-controller in bare-metal environment, please make sure both the proxy service and admin api service of Apache APISIX are exposed outside of the Kubernetes cluster, e.g. configuring them as [NodePort](https://kubernetes.io/docs/concepts/services-networking/service/#nodeport) services.
+
+Let's assume the Admin API service address of Apache APISIX is `http://192.168.65.2:31156`. Next launch the ingress-apisix-controller by the following command.
+
+```shell
+cd /path/to/apisix-ingress-controller
+./apisix-ingress-controller ingress \
+    --kubeconfig /path/to/kubeconfig \
+    --http-listen :8080 \
+    --log-output stderr \
+    --default-apisix-cluster-base-url http://192.168.65.2:31156/apisix/admin
+    --default-apisix-cluster-admin-key edd1c9f034335f136f87ad84b625c8f1
+```
+
+Something you need to pay attention to:
+
+* configuring of `--kubeconfig`, if you are using Minikube, the file path should be `~/.kube/config`.
+* configuring of `--default-apisix-cluster-admin-key`, if you have changed the admin key in Apache APISIX, also changing it here. If you have disabled the authentication in Apache APISIX, just removing this option.
+
+## Pre-commit todo
+
+When everything is ready, before submitting the code, please make sure that the license, code style, and document format are consistent with the project specification.
+
+We provide commands to implement it, just run the following commands:
+
+```shell
+make update-codegen
+make update-license
+make update-gofmt
+make update-mdlint
+```
+
+or just run one command:
+
+```shell
+make update-all
+```
 
 ## How to add a new feature or change an existing one
```

Comment:
I think it would be better to put this section before `test`.

Reply:
Line 122 to 124 or the paragraph Pre-commit todo?

Reply:
I mean the "How to add a new feature or change an existing one" section, sorry.